hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -763,47 +842,130 @@ object PruneDeadFields {
           memoizeValueIR(right, rType.copy(elementType = rRequested), memo))
 
       case ArrayFor(a, valueName, body) =>
         assert(requestedType == TVoid)
         val aType = a.typ.asInstanceOf[TArray]
         val bodyEnv = memoizeValueIR(body, body.typ, memo)
-        val valueType = bodyEnv.lookupOption(valueName).map(_._2).getOrElse(minimal(-aType.elementType))
+        val valueType = unifySeq(
+          aType.elementType,
+          bodyEnv.eval.lookupOption(valueName).map(_.result()).getOrElse(Array()))
         unifyEnvs(
-          bodyEnv.delete(valueName),
+          bodyEnv.deleteEval(valueName),
           memoizeValueIR(a, aType.copy(elementType = valueType), memo)
         )
-      case AggExplode(a, name, body, _) =>
+      case AggExplode(a, name, body, isScan) =>
         val aType = a.typ.asInstanceOf[TArray]
-        val bodyDep = memoizeValueIR(body,
+        val bodyEnv = memoizeValueIR(body,
           requestedType,
           memo)
-        val valueType = bodyDep.lookupOption(name).map(_._2).getOrElse(minimal(-aType.elementType))
+        if (isScan) {
+          val valueType = unifySeq(
+            aType.elementType,
+            bodyEnv.scanOrEmpty.lookupOption(name).map(_.result()).getOrElse(Array()))
+
+          val aEnv = memoizeValueIR(a, aType.copy(elementType = valueType), memo)
+          unifyEnvs(
+            BindingEnv(scan = bodyEnv.scan.map(_.delete(name))),
+            BindingEnv(scan = Some(aEnv.eval))
+          )
+        } else {
+          val valueType = unifySeq(
+            aType.elementType,
+            bodyEnv.aggOrEmpty.lookupOption(name).map(_.result()).getOrElse(Array()))
+
+          val aEnv = memoizeValueIR(a, aType.copy(elementType = valueType), memo)
+          unifyEnvs(
+            BindingEnv(agg = bodyEnv.agg.map(_.delete(name))),
+            BindingEnv(agg = Some(aEnv.eval))
+          )
+        }
+      case AggFilter(cond, aggIR, isScan) =>
+        val condEnv = memoizeValueIR(cond, cond.typ, memo)
         unifyEnvs(
-          bodyDep.delete(name),
-          memoizeValueIR(a, aType.copy(elementType = valueType), memo)
+          if (isScan)
+            BindingEnv(scan = Some(condEnv.eval))
+          else
+            BindingEnv(agg = Some(condEnv.eval)),
+          memoizeValueIR(aggIR, requestedType, memo)
         )
-      case AggArrayPerElement(a, name, aggBody, _) =>
+      case AggGroupBy(key, aggIR, isScan) =>
+        val keyEnv = memoizeValueIR(key, key.typ, memo)
+        unifyEnvs(
+          if (isScan)
+            BindingEnv(scan = Some(keyEnv.eval))
+          else
+            BindingEnv(agg = Some(keyEnv.eval)),
+          memoizeValueIR(aggIR, requestedType, memo)
+        )
+      case AggArrayPerElement(a, name, aggBody, isScan) =>
         val aType = a.typ.asInstanceOf[TArray]
-        val bodyDep = memoizeValueIR(aggBody,
+        val bodyEnv = memoizeValueIR(aggBody,
           requestedType.asInstanceOf[TArray].elementType,
```

Comment:
how is this somehow getting a TDict as the requested type? Python tests are failing here, I believe?

Reply:
yep debugging, found some other stuff too (like no memoize rule for ArraySort, which led to free variables getting propagated up -- this wasn't a bug until I changed the prune structure)